### ISSUE TITLE
GitHub Pages + Jekyll + Swagger = Pretty API Spec

### ISSUE THREAD

**Issue by BranicYeti** [Zyncco/Website#2]
```text
Vilsol, [07.03.17 14:17]
    can we please use something other than GIT to organize the fucking API
Vilsol, [07.03.17 14:17]
    so PITA to navigate
Vilsol, [07.03.17 14:18]
    no
Vilsol, [07.03.17 14:18]
    like
Vilsol, [07.03.17 14:18]
    use GIT
Vilsol, [07.03.17 14:19]
    but use some visual html generator
Brandon Currell, [07.03.17 14:20]
    As in like Doxygen or Sphinx, or gh-pages?
Vilsol, [07.03.17 14:21]
    gh-pages, but im choosing a generator atm
Vilsol, [07.03.17 14:21]
    Swagger looks pretty good
Vilsol, [07.03.17 14:24]
    Brandon try http://editor.swagger.io/
Vilsol, [07.03.17 14:25]
    its really nice
```

I agree with him, Swagger is nice. Here's my idea.

We use our static site (`zync.co`) as more than just a landing page. We also host a pretty HTML version of our API spec on it using a mixture of Jekyll and Swagger. All of the work is already done for us (see links below), we just need to create the Swagger JSON file. I do not think that this should be an alternative to the Git repo with all of the Markdown files though. I think we should provide both. It wouldn't be too difficult to manage both sets of data since Swagger will take a single JSON file with all of the endpoints and information in it. We might even be able to create a script that populates the Swagger JSON file for us from the Markdown, or vice-versa, but in my opinion, that won't be necessary since we shouldn't need to change the Spec much once it's finished.

Thoughts?

---

https://help.github.com/categories/customizing-github-pages/
https://jexhson.github.io/jekyll-swagger/

**Comment by Vilsol** [comment]
I think it would be easier to generate the markdown Doc's from swagger, as it already has a similar functionality and it's easy to make plugins for it

**Comment by BranicYeti** [comment]
I'm not sure that we can run Swagger plugins within gh-pages, but i haven't looked extensively yet. I'm trying to reduce our server requirements. With this current setup, we don't need one at all, hence why it might seem a bit more complicated than running Swagger on a webserver. 

My main worry with Swagger -> Markdown is that it might not break the endpoints into separate directories like the Clipboard example i uploaded in the initial commit. I thought that might be better for reasons relating to bandwidth and page load times. That being said, it's probably easier to parse YAML with a script than it is to manually parse Markdown. Depending on who will work on this sort of thing, it might be best to save this conversation for the meeting tomorrow.

**Comment by BranicYeti** [comment]
```text
Vilsol [5:02 PM] 
    OK, we can't use swagger for API docs
    I'm not sure why any company ever would
    The entire doc HAS to be a single file
```

Alternative suggested: [RAML](http://raml.org/)

By the looks of it, there isn't an existing component that would allow us to use GitHub pages to display a pretty API spec automatically, just by dropping it in. However, it could be developed. The pretty API spec via GitHub pages just got pushed down to a lower priority unless we host it on a proper webserver.

On a site note, I believe Swagger 2.0 included changes that allow API designers to break the spec into multiple files, however more information is to be found on that regarding whether or not it's efficient / effective to do so.

**Comment by Vilsol** [comment]
RAML has a generator for HTML, Markdown and even Laravel, it fully supports gh-pages.

**Comment by aaomidi** [comment]
Great progress guys!

**Comment by BranicYeti** [comment]
I think we have decided to make the switch to RAML. To get a pretty HTML version of the API spec, we will use Travis to "build" the HTML from RAML using the provided tool `raml2html`. These generated pages will be pushed to a branch within the Spec repo from which GitHub pages will run on the domain `docs.zync.co` or similar. People should probably start reading up on the structure of RAML directories and files. It seems rather intuitive, so good luck.

Closing this issue. A new one should be created on the Spec repo if it is required.
